### ISSUE TITLE
Many packages missing PHP extensions in "require", "require-dev" or "suggest"

### ISSUE THREAD

**Issue by vbashkirtsev** [composer/composer#5725]
Earlier today I have proposed a PR #5720 which was quickly closed off as @alcohol was afraid that accepting it would open floodgate for promotion of other plugins (actually the tool I proposed is not a plugin but rather standalone tool).

But the issue I was trying to address still remains. Many packages on Packagist do not list their dependencies on PHP extensions or these dependencies are incomplete causing a bit of hassles when deploying dependencies on fresh minimal install system. Generally minimal install (in my case FC23 but I am sure it is valid for other systems) missing ext-xml ext-dom ext-mbstring ext-soap ext-mysqli and number of other commonly used extensions. In fact composer installation of many packages succeed on minimal install systems just to fail in run time with fatal errors. I personally do a bit of deployment this way and find it extremely annoying and feel insecure that there may be some dependency which uses some PHP extension I don't have installed and it does not manifest a problem straight away and may blow up in the customer face at any point later.

So I looked around for the solution, asked question on SO site, found nothing and so made my own tool which runs through the code, analyzes it, checks it against composer.json and reports which extensions used by code not mentioned in composer.json "require", "require-dev" or "suggest" sections. I have run it against number of well known packages and was really shocked that many of them in fact missing extensions dependencies. For example monolog/monolog 1.21.0 (which is mentioned on getcomposer.org):

```
Following extensions are missing in require section of composer.json:

  ext-json
    Used json_encode in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/FlowdockHandler.php:108
    Used json_encode in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Handler/ChromePHPHandlerTest.php:36
    Used json_encode in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Handler/ChromePHPHandlerTest.php:60
    Used json_encode in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Handler/ChromePHPHandlerTest.php:103

  ext-soap
    Used SoapFault in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Formatter/NormalizerFormatter.php:130
    Used SoapFault in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Formatter/NormalizerFormatterTest.php:95

  ext-xml
    Used utf8_encode in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/ChromePHPHandler.php:138
    Used utf8_encode in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/ChromePHPHandler.php:153
    Used utf8_encode in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Handler/ChromePHPHandlerTest.php:36
    Used utf8_encode in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Handler/ChromePHPHandlerTest.php:60
    Used utf8_encode in /home/mega/monolog/vendor/monolog/monolog/tests/Monolog/Handler/ChromePHPHandlerTest.php:103

  ext-mbstring
    Used mb_substr in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/HipChatHandler.php:148
    Used mb_strlen in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/HipChatHandler.php:345

  ext-hash
    Used hash in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Processor/UidProcessor.php:29

  ext-sockets
    Used socket_create in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/CubeHandler.php:71
    Used AF_INET in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/CubeHandler.php:71
    Used SOCK_DGRAM in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/CubeHandler.php:71
    Used socket_connect in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/CubeHandler.php:76
    Used socket_send in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/CubeHandler.php:134

  ext-redis
    Used Redis in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/RedisHandler.php:43
    Used Redis in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/RedisHandler.php:75

  ext-curl
    Used curl_init in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:87
    Used curl_setopt in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:89
    Used CURLOPT_URL in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:89
    Used curl_setopt in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:90
    Used CURLOPT_POST in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:90
    Used curl_setopt in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:91
    Used CURLOPT_POSTFIELDS in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:91
    Used curl_setopt in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:92
    Used CURLOPT_HTTPHEADER in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:92
    Used curl_setopt in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:93
    Used CURLOPT_RETURNTRANSFER in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/LogglyHandler.php:93

  ext-filter
    Used filter_var in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/SocketHandler.php:240
    Used FILTER_VALIDATE_FLOAT in /home/mega/monolog/vendor/monolog/monolog/src/Monolog/Handler/SocketHandler.php:240

```

Checking doctrine/annotations:

`Following extensions are missing in require section of composer.json: ext-tokenizer ext-ctype ext-json`

Checking fabpot/php-cs-fixer:

`Following extensions are missing in require section of composer.json: ext-Phar ext-json ext-xml ext-dom`

... and this list can go on.

I believe @alcohol has valid point about opening floodgate for promotion of plugins/etc but in this case I would like to promote best practice and not the tool itself - I just believe it is reasonable to mention it as it was purpose built for this reason and can make Composer/Packagist ecosystem healthier.

One way or another the issue of missing PHP dependencies should be addressed. Be it with my tool or with any other (including such functionality into Composer itself - but I believe it is not best fit).

**Comment by alcohol** [comment]
If you want to introduce it to the community, there are better channels.

* http://reddit.com/r/php
* irc://freenode.net/php
* Your own blog
* twitter

etc..

**Comment by Seldaek** [comment]
IMO, the reason most packages don't require those extensions like ext-tokenizer ext-ctype ext-json is that they are enabled by default, and if you compile php with --disable-all you are just asking for trouble really. It causes pain for everyone, and is kinda pointless. If we can't rely on a base set of functions as being part of the language, everyone suffers. So yeah these are not generally seen as an additional requirement but rather just being part of the php requirement. Adding your tool to the docs is not gonna change that fact.

**Comment by vbashkirtsev** [comment]
In this post I even did not mention my tool by name. I was trying to stress that there's widespread problem which should be addressed. And it is not about promotion of my tool.

@stof: it is not required to put extension dependency into "require" - it may go into "suggest" section as well (and some well known packages do this way). I personally would appreciate to see "monolog/monolog suggests installing ext-redis" message during installation and know that it may be used.

@alcohol: this is issue not about introduction of any particular tool - it is about issue of missing extensions on minimal install systems and composer being unable to catch it (at no composer's fault).

@Seldaek: On FC23 default PHP install missing ext-xml ext-dom ext-mysqli ext-mbstring - these extensions are commonly used by many packages. I personally hit this issue many times and sure there's plenty of people wasting their own time for nothing the same way. And at this point it is not about adding my tool to the docs: I made PR, it was closed off, reason explained, moving on. It is about doing something to prevent this issue to proliferate further.

**Comment by Seldaek** [comment]
We do suggest a bunch of stuff in monolog https://github.com/Seldaek/monolog/blob/master/composer.json#L33 - ext-redis and a few others are missing, because we are human and make mistakes, feel free to send PRs there :)

Aside from that, yes maybe some extensions like xml are less standard, I don't know if Fedora is the outlier or not there. I'd recommend just sending PRs to the offending projects when you spot missing requirements, because that's the best way to move things forward and get people to notice. I don't really see what composer can do as a project.

**Comment by vbashkirtsev** [comment]
@Seldaek: Totally agree with you - generally missing extensions happen due to human nature. I used my own tool on my own packages and was really depressed how not diligent I was :(

To the moment I did so (for example phpunit was missing few - Sebastian accepted PR to add dependencies with no problems). But such effort would not be enough to fix millions of packages out there. I also don't see it as part of composer but still whole Composer/Packagist ecosystem would benefit if something is done about it. May be it may be feasible to create a separate tool which helps to manage composer.json which will take care of the issue - I'll leave it up to you.

Also at one point someone, who actively uses composer, expressed their surprise that PHP extensions may be depend upon in composer.json . May be section about extensions dependency of composer's docs should be elaborated a bit more so people would notice such functionality.
